sources/Algo/src/Algo.cpp

```diff
@@ -81,16 +83,53 @@ GeneratorDefinitionAlgorithm::operator()(const NodePtr& node) {
   if (node_generators.size() == 1) {
     auto model = useInfiniteReactivelimits_ ? GeneratorDefinition::ModelType::SIGNALN : GeneratorDefinition::ModelType::DIAGRAM_PQ_SIGNALN;
     const auto& gen = node_generators.front();
+
+    if (model == GeneratorDefinition::ModelType::DIAGRAM_PQ_SIGNALN && !isDiagramValid(gen)) {
+      return;
+    }
+
     generators_.emplace_back(gen.id, model, node->id, gen.points, gen.qmin, gen.qmax, gen.pmin, gen.pmax);
   } else {
     auto model =
         useInfiniteReactivelimits_ ? GeneratorDefinition::ModelType::WITH_IMPEDANCE_SIGNALN : GeneratorDefinition::ModelType::WITH_IMPEDANCE_DIAGRAM_PQ_SIGNALN;
     for (auto it = node_generators.begin(); it != node_generators.end(); ++it) {
+      if (model == GeneratorDefinition::ModelType::WITH_IMPEDANCE_DIAGRAM_PQ_SIGNALN && !isDiagramValid(*it)) {
+        continue;
+      }
       generators_.emplace_back(it->id, model, node->id, it->points, it->qmin, it->qmax, it->pmin, it->pmax);
     }
   }
 }
 
+bool
+GeneratorDefinitionAlgorithm::isDiagramValid(const inputs::Generator& generator) {
+  bool valid = true;
+
+  if (generator.points.size() == 0) {
+    return true;
+  }
+  if (generator.points.size() == 1) {
+    LOG(warn) << MESS(InvalidDiagram, generator.id) << LOG_ENDL;
+    return false;
+  }
+
+  auto firstP = generator.points.front().p;
+  bool allQminEqualQmax = true;
+  bool allPEqual = true;
+  auto it = generator.points.begin();
+  while ((allQminEqualQmax || allPEqual) && it != generator.points.end()) {
```

Comment:
I always prefer for loops to while loops

Reply:
I prefer the while loops, it isn't mandatory to have for loops ?

Reply:
Nope, not mandatory, just more elegant

Reply:
It boils down to personal preference

Reply:
absolutely :)

Reply:
Just to explain my preference: the main interest of for loop is to avoid having to handle the iterator inside the loop, which can be dangerous if the complexity increase. Here the algorithm is simple so it's ok to keep it that way if you prefer